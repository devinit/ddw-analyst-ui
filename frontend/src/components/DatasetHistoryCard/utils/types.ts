export interface SavedQueryData {
  id: number;
  saved_query_db_table: string;
  operation: number;
  full_query: string | null;
  active: boolean;
  description: string;
  created_on: string;
  user: string;
  status: string;
}
