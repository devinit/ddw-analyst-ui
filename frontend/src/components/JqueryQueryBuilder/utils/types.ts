export type JqueryQueryBuilderFilter = {
  condition?: string;
  rules?: (JqueryQueryBuilderFilterComparator | JqueryQueryBuilderFilter)[];
};

export type JqueryQueryBuilderHaving = {
  condition?: string;
  rules?: (JqueryQueryBuilderHavingComparator | JqueryQueryBuilderHaving)[];
};

export type JqueryQueryBuilderHavingComparator = {
  id: string;
  field: string;
  type: string;
  input: string;
  operator: JqueryQueryBuilderComps;
  value?: string | number | { plain: string | number } | { column: string; aggregate: string };
  values?: { value: string; label: string }[];
};

export type JqueryQueryBuilderFilterComparator = {
  id: string;
  field: string;
  type: string;
  input: string;
  operator: JqueryQueryBuilderComps;
  value: string | number;
};
export type JqueryQueryBuilderComps =
  | 'equal'
  | 'not_equal'
  | 'less'
  | 'less_or_equal'
  | 'greater'
  | 'greater_or_equal'
  | 'between'
  | 'in';

export type JqueryQueryBuilderFieldData = {
  id: string | number;
  label: string;
  type: string;
};

export type JqueryQueryBuilderIcons = {
  add_group: string;
  add_rule: string;
  remove_group: string;
  remove_rule: string;
  error: string;
};
