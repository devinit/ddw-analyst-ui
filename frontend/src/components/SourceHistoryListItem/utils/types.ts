export interface FrozenData {
  id: number;
  parent_db_table: string;
  frozen_db_table: string | null;
  active: boolean;
  description: string;
  created_on: string;
  user: string;
  status: string;
  logs: string | null;
}
