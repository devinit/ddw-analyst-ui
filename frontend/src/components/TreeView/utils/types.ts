import { ReactElement } from 'react';

export interface TreeViewProps {
  data: Data[];
  depth?: number;

  deleteElement?: ReactElement;

  getStyleClassCb?: (node: Data) => void;

  isCheckable?: (node: Data, depth?: number) => void;
  isDeletable?: (node: Data, depth?: number) => void;
  isExpandable?: (node: Data, depth?: number) => void;

  keywordChildren?: string;
  keywordChildrenLoading?: string;
  keywordKey?: string;
  keywordLabel?: string;

  loadingElement?: ReactElement;
  noChildrenAvailableMessage?: string;

  onCheckToggleCb?: (data: Data[], depth?: number) => void;
  onDeleteCb?: (node: Data, data: Data[], depth?: number) => void;
  onExpandToggleCb?: (node: Data, depth?: number) => void;
  onUpdateCb?: (data: unknown[], depth?: number) => void;

  transitionEnterTimeout?: number;
  transitionExitTimeout?: number;
}

export interface Data {
  id: number;
  name: string;
  isExpanded?: boolean;
  isChecked?: boolean;
  isChildrenLoading?: boolean;
  children?: Data[];
}
