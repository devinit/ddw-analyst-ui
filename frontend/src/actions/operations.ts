import {
  FETCH_OPERATION,
  FETCH_OPERATIONS,
  FETCH_OPERATION_FAILED,
  OperationsAction,
  SET_ACTIVE_OPERATION
} from '../reducers/operations';
import { OperationMap } from '../types/operations';

interface FetchOptions {
  limit?: number;
  offset?: number;
  link?: string;
}
export const fetchOperations = ({ limit = 10, offset = 0, link }: FetchOptions): Partial<OperationsAction> =>
  ({ type: FETCH_OPERATIONS, payload: { limit, offset, link } });
export const fetchOperation = (id: string): Partial<OperationsAction> => ({ type: FETCH_OPERATION, payload: { id } });
export const setOperation = (activeOperation?: OperationMap, loading = false): Partial<OperationsAction> =>
  ({ type: SET_ACTIVE_OPERATION, activeOperation, loading });
export const fetchOperationFailed = (): Partial<OperationsAction> => ({ type: FETCH_OPERATION_FAILED });
