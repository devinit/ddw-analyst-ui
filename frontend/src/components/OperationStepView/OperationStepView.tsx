import React, { FunctionComponent } from 'react';
import { Badge, Button } from 'react-bootstrap';
import { Popup } from 'semantic-ui-react';
import styled from 'styled-components';
import { OperationStepMap } from '../../types/operations';

interface OperationStepProps {
  step: OperationStepMap;
  onDuplicateStep: (step?: OperationStepMap) => void;
}

const StyledPopContent = styled.div`
  font-size: 0.8rem;
`;
const StyledPopup = styled(Popup)`
  &.ui.popup {
    border: none;
  }
`;
const StyledIcon = styled.i`
  font-size: 18px;
  margin-right: 5px;
`;
const StyledButton = styled(Button)`
  padding-top: 0 !important;
  padding-right: 0 !important;
  margin-top: 0;
`;
const StyledContentWrapper = styled.div`
  position: relative;
`;
const StyledActionsWrapper = styled.div`
  position: absolute;
  display: inline-block;
  right: 0;
  margin-right: 1rem;
`;

export const OperationStepView: FunctionComponent<OperationStepProps> = ({ step, ...props }) => {
  const onDuplicate = (event: React.MouseEvent<HTMLDivElement>) => {
    event.preventDefault();
    event.stopPropagation();
    props.onDuplicateStep(step);
  };
  const renderStepInfo = (step: OperationStepMap) => {
    return (
      <StyledActionsWrapper>
        {step.get('description') ? (
          <StyledPopup
            trigger={
              <StyledIcon className="material-icons mt-1" data-testid="step-info-trigger">
                info
              </StyledIcon>
            }
            content={<StyledPopContent>{step.get('description')}</StyledPopContent>}
            basic
          />
        ) : null}
        <StyledButton
          className="pl-1"
          title="Duplicate"
          variant="link"
          size="sm"
          data-testid="step-duplicate"
          onClick={onDuplicate}
        >
          <StyledIcon className="material-icons" data-testid="step-duplicate-trigger">
            content_copy
          </StyledIcon>
        </StyledButton>
      </StyledActionsWrapper>
    );
  };

  return (
    <React.Fragment>
      <div>
        <Badge variant="secondary">
          {(step.get('query_func') as string).toUpperCase().split('_').join(' ')}
        </Badge>
      </div>
      <StyledContentWrapper>
        {step.get('name')}
        {renderStepInfo(step)}
      </StyledContentWrapper>
    </React.Fragment>
  );
};
