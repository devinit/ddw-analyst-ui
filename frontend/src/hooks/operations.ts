import axios, { AxiosResponse } from 'axios';
import { fromJS } from 'immutable';
import * as localForage from 'localforage';
import { useEffect, useState } from 'react';
import { Operation, OperationMap } from '../types/operations';
import { api, localForageKeys } from '../utils';

interface UseOperationResult {
  operation?: OperationMap;
  loading: boolean;
}

export const useOperation = (id?: number, fetch = false): UseOperationResult => {
  if (!id) {
    return { loading: false, operation: undefined };
  }
  const [operation, setOperation] = useState<OperationMap | undefined>(fromJS({}));
  const [loading, setLoading] = useState(false);

  const fetchOperation = () => {
    setLoading(true);
    axios
      .request({
        url: `${api.routes.SINGLE_DATASET}${id}`,
        method: 'get',
        headers: {
          'Content-Type': 'application/json',
        },
      })
      .then(({ status, data, statusText }: AxiosResponse<Operation>) => {
        if (status === 200 && data) {
          const activeOperation = fromJS(data);
          setOperation(activeOperation);
          setLoading(false);
        } else if (status === 401) {
          console.log('Failed to fetch operation: ', statusText);
          setOperation(undefined);
          setLoading(false);
        }
      })
      .catch((error) => {
        console.log(
          `Failed to fetch operation: ${error.response.status} ${error.response.statusText}`,
        );
        setOperation(undefined);
        setLoading(false);
      });
  };

  useEffect(() => {
    if (fetch) {
      fetchOperation();
    } else {
      localForage
        .getItem<Operation | undefined>(localForageKeys.ACTIVE_OPERATION)
        .then((activeOperation) => {
          if (activeOperation && activeOperation.id === id) {
            setOperation(fromJS(activeOperation));
            setLoading(false);
          } else {
            fetchOperation();
          }
        });
    }
  }, [id]);

  return { loading, operation };
};
