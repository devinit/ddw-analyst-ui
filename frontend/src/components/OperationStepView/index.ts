import { OperationStepView } from './OperationStepView';

export * from './OperationStepView';
export { OperationStepView as default };
