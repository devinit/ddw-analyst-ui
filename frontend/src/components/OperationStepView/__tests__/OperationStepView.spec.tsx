/**
 * @jest-environment jsdom
 */
import { fireEvent, render, screen } from '@testing-library/react';
import { Map } from 'immutable';
import 'jest-styled-components';
import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import { OperationStep, OperationStepMap } from '../../../types/operations';
import { OperationStepView } from '../OperationStepView';

const step: Partial<OperationStep> = {
  name: 'My Test Step',
  updated_on: new Date('2020-01-02').toISOString(),
  query_func: 'filter',
  query_kwargs: 'kwargs',
};

const props = {
  step: Map(step) as OperationStepMap,
  onDuplicateStep: jest.fn(),
};

describe('OperationStep', () => {
  afterEach(() => {
    props.step = Map(step) as OperationStepMap;
  });

  test('renders correctly with the default props', () => {
    const renderer = TestRenderer.create(<OperationStepView {...props} />).toJSON();

    expect(renderer).toMatchSnapshot();
  });

  test('renders an info button when the provided step has a description', () => {
    props.step = Map({ ...step, description: 'My Step Description' }) as OperationStepMap;
    const renderer = TestRenderer.create(<OperationStepView {...props} />).toJSON();

    expect(renderer).toMatchSnapshot();
  });

  test('renders a copy button that responds to clicks', () => {
    render(<OperationStepView {...props} />);
    const copyButton = screen.getByTestId('step-duplicate');
    fireEvent.click(copyButton);

    expect(props.onDuplicateStep).toHaveBeenCalled();
  });
});
