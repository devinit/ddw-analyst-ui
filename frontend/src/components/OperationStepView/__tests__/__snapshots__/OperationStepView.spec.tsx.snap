// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OperationStep renders an info button when the provided step has a description 1`] = `
Array [
  <div>
    <span
      className="badge bg-secondary"
    >
      FILTER
    </span>
  </div>,
  .c2 {
  font-size: 18px;
  margin-right: 5px;
}

.c3 {
  padding-top: 0 !important;
  padding-right: 0 !important;
  margin-top: 0;
}

.c0 {
  position: relative;
}

.c1 {
  position: absolute;
  display: inline-block;
  right: 0;
  margin-right: 1rem;
}

<div
    className="c0"
  >
    My Test Step
    <div
      className="c1"
    >
      <i
        className="c2 material-icons mt-1"
        data-testid="step-info-trigger"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        info
      </i>
      <button
        className="c3 pl-1 btn btn-link btn-sm"
        data-testid="step-duplicate"
        disabled={false}
        onClick={[Function]}
        title="Duplicate"
        type="button"
      >
        <i
          className="c2 material-icons"
          data-testid="step-duplicate-trigger"
        >
          content_copy
        </i>
      </button>
    </div>
  </div>,
]
`;

exports[`OperationStep renders correctly with the default props 1`] = `
Array [
  <div>
    <span
      className="badge bg-secondary"
    >
      FILTER
    </span>
  </div>,
  .c3 {
  font-size: 18px;
  margin-right: 5px;
}

.c2 {
  padding-top: 0 !important;
  padding-right: 0 !important;
  margin-top: 0;
}

.c0 {
  position: relative;
}

.c1 {
  position: absolute;
  display: inline-block;
  right: 0;
  margin-right: 1rem;
}

<div
    className="c0"
  >
    My Test Step
    <div
      className="c1"
    >
      <button
        className="c2 pl-1 btn btn-link btn-sm"
        data-testid="step-duplicate"
        disabled={false}
        onClick={[Function]}
        title="Duplicate"
        type="button"
      >
        <i
          className="c3 material-icons"
          data-testid="step-duplicate-trigger"
        >
          content_copy
        </i>
      </button>
    </div>
  </div>,
]
`;
